### ISSUE TITLE
Feature proposal: *spaces*

### ISSUE THREAD

**Issue by brisvag** [napari/napari#4419]
### The problem
- Often datasets are composed of many different images that do not share a coordinate space (i.e: no reason/way to spatially or temporally relate them along a dimension); here I'm calling these separate world "*spaces*".
- One often wants to have access to all the *spaces* without creating/deleting layers every time (and reading from disk every time)
- It's often useful to quickly switch between *spaces* to compare features between samples.

### Current approaches
1. simply load all the layers in the layerlist. This has 3 major downsides:
    - you have to toggle layer visibility to switch between *spaces*, which is a nightmare (especially if multiple layers share the same space, and linking them is out of question). This can be simplified with a widget (I currently [do that in blik](https://github.com/gutsche-lab/blik/blob/e175659d212b5a96d6868668aa780286ce7a068c/blik/widgets/volume_selector.py#L51-L71)) but it's not ideal.
    - the layerlist is basically useless once many layers are loaded, cause it's impossible to find what you need and gives no useful visual cuues to the state of the viewer. (*this last part will be partially solved by [layer groups](https://github.com/napari/napari/pull/4018), but not the general problem of having many __spaces__*)
    - If you have many layers (> hundreds), opening the whole dataset can take **minutes**. Time taken is [no longer exponential](https://github.com/napari/napari/pull/1945), but still unusable in case of large datasets. This probably won't be addressed by the async efforts, because the issue is not with data loading but with vispy node creation and hooking up and firing of events (afaik).
 2. create a widget that creates/inserts layers into the layer list on demand. This used to be my previous solution in `blik`.
    - does not have the problems above, because only the layers that are actually viewed are added to the viewer, making it easier to see what's what and faster to load
    - however, it's not a napari native solution, which means that
        - we can't use reader plugins to create the layers because they woudl all be added at once
        - the user/plugin is forced to keep track of everything that does not stay in the viewer (which means holding a reference to some sort of "dataset" object
        - bye bye serialization

### The Solution
Make some form of point `2.` a first class citizen in napari.
- The viewer should be aware of *spaces* (or workspaces/scenes/worlds/whatever) which are collections of layers that are dimensionally unrelated to one-another but live in the same viewer
- Make it easy to see what *spaces* exist and to switch between them

Potentially, once layer groups are a thing, *spaces* could simply be the nodes at depth 1 in the layer graph.

**Comment by MosGeo** [comment]
This would be good for for datasets with different modality and scales, e.g., if you have 3D CT scan and SEM of the same sample.

I also see this as a precursor to a 'project' system.

**Comment by jni** [comment]
> bye bye serialization

:rofl: 

Nice issue @brisvag!

Although I see the need, I wonder why you would do this vs multiple viewers? Then spaces are serialisations of that particular viewer?

It would be good to have a *concrete* example where you have multiple spaces *in one file*, for example.

I also do see layergroups as a prerequisite here.

**Comment by brisvag** [comment]
> Although I see the need, I wonder why you would do this vs multiple viewers? Then spaces are serialisations of that particular viewer?
> 
> It would be good to have a _concrete_ example where you have multiple spaces _in one file_, for example.
> 
> I also do see layergroups as a prerequisite here.

My use case is *hundreds* of spaces. Here's an example where I'm using trick number `1.` (toggle visibility through a widget). In this example I only have one image per space, but If I had particle picking data I would have 4 more layers per space. I think havign that many viewers might be counterproductive :P Also, I want the data to be accessible all from the same place.

https://user-images.githubusercontent.com/23482191/165916904-93abe064-cee1-4c87-9efb-d41dbc586899.mp4

I don't think layergroups are a prerequisite; rather an enhancement. They can work together, and they solve slightly different problems. What about my proposal above you think requires layegroups?

**Comment by alisterburt** [comment]
> If you have many layers (> hundreds), opening the whole dataset can take minutes. Time taken is https://github.com/napari/napari/pull/1945, but still unusable in case of large datasets. This probably won't be addressed by the async efforts, because the issue is not with data loading but with vispy node creation and hooking up and firing of events (afaik).

Are vispy nodes only created/attached when layers are added to the layerlist? nice!

> I don't think layergroups are a prerequisite; rather an enhancement. They can work together, and they solve slightly different problems. What about my proposal above you think requires layegroups?

I agree here - this feels separate issues to layer groups, high value added for potentially low effort?

As an API, how about something like?

```
Viewer.spaces: EventedDict[LayerList]
Viewer.active_space: str
```

where `Viewer.active_space` indexes `viewer.spaces`

**Comment by alisterburt** [comment]
super basic prototype on a plugin widget

https://github.com/alisterburt/napari-spaces/blob/fc4840aaec2265af427a6600f0a35de94070feaa/src/napari_spaces/_widget.py#L16-L50

https://user-images.githubusercontent.com/7307488/166587193-6c5e1417-93a9-4a53-8709-c3308d9ad1ff.mp4

potentially a better API would be:


```python
class Space(BaseModel):
    name: str
    layers: List[Layer]

Viewer.spaces: SelectableEventedList[Space]

def replace_layers():
    viewer.layers.clear()
    viewer.layers.extend(viewer.spaces.selection.active)

viewer.spaces.selection.events.active.connect(replace_layers)
```

**Comment by alisterburt** [comment]
cc @junxini this is the 'dataset navigation' we discussed - the UI is obviously not great and should probably be replaced by some of @Czaki's work over at napari/superqt#80

It could be that 'workflows' are defined in one space then applied over many spaces

**Comment by junxini** [comment]
> cc @junxini this is the 'dataset navigation' we discussed - the UI is obviously not great and should probably be replaced by some of @Czaki's work over at [napari/superqt#80](https://github.com/napari/superqt/pull/80)

thank you for tagging me Alister, seeing the prototype really helps bring it to life!
> It could be that 'workflows' are defined in one space then applied over many spaces

I imagined 'workflows' would be applied across multiple images/files and can definitely see a use case for having it applied across multiple spaces. Excited to see how spaces will improve the user experience

**Comment by alisterburt** [comment]
> I imagined 'workflows' would be applied across multiple images/files and can definitely see a use case for having it applied across multiple spaces. Excited to see how spaces will improve the user experience

totally - there are lots of potential use cases where the data in a 'space' is not well represented by a single file on disk. I think having workflows start from layers in napari spaces is probably more flexible than assuming batch processing will always work over a set of files 🙂 

This would separate the 'batch processing' idea you're thinking about into
- develop simple ways to load batches of data into spaces
- develop the infrastructure for executing workflows on a per space basis

the UX would be quite nice I think, mirrors quite nicely what the deep learning community have done in terms of separation of concerns (pytorch `DataLoader`s are totally separate from the part which actually does the computation)

Let's keep this line of thought open 🙂

**Comment by Czaki** [comment]
In the exception of loading time, I still do not understand why layer groups do not solve this problem? There is no problem that the writer could read multiple files from the disc, and return single layer groups. 

Maybe above some number of layers, the layer list should have a search box? and the option Show this with hiding all others? 

Or do you mean about situations where you would like to use a single layer in multiple spaces?

**Comment by alisterburt** [comment]
@Czaki I totally agree that layergroups can solve this problem - getting the GUI right feels really important though. Building on what you said, what do you think of the proposal below

<img width="518" alt="image" src="https://user-images.githubusercontent.com/7307488/168499909-d180f27f-6ae4-426e-ad5c-a90ace7e9425.png">

In the image above, the root nodes of the layer group tree are the 'spaces', navigated with your combo/search box. Only one of these root nodes can be visible at a time. The layers in each 'space' (branch of the tree) are controlled with controls as normal

**Comment by Czaki** [comment]
So spaces will be a list of possible views for fast switching between them? 

Something like multiple files panel from PartSeg https://partseg.readthedocs.io/en/latest/interface-overview/interface-overview.html#multiple-files-panel but with search (and maybe a better option for renaming elements?

**Comment by brisvag** [comment]
While layer groups will help *in general* they don't actually address the same problem. My proposal for spaces is intended to solve the problem of **data that does not belong to the same coordinates system**. I think [this comment above](https://github.com/napari/napari/issues/4419#issuecomment-1113090992) explains well why layergroups are not a solution. Indeed, `PartSeg`'s file panel is basically targeting the same issue, but would be limited to *files*. We can generalize that to any collection of layers.

> So spaces will be a list of possible views for fast switching between them?

I guess... which makes me think: maybe we should also allow for a layer to exist in multiple spaces at once? This would allow people to use spaces in a different way: to quickly load "presets" of combinations of layers in order to look at different visualizations. This would play well into our future efforts towards multi-canvas (you could simply activate more than one space, and boom).

**Comment by Czaki** [comment]
> Indeed, `PartSeg`'s file panel is basically targeting the same issue, but would be limited to _files_. We can generalize that to any collection of layers

Of course, this cannot be used directly because it is pinned to the inner PartSeg data representation (which is still different from napari even after some migrations). I would link it as an example of an idea/interface. 

> but would be limited to _files_
PartSeg uses a single entry point file load so a single entry could represent multiple files. Also, the code of the widget does not assume that it needs to be a real file path but a unique name. 

> My proposal for spaces is intended to solve the problem of **data that does not belong to the same coordinates system**. I think [this comment above](https://github.com/napari/napari/issues/4419#issuecomment-1113090992) explains well why layergroups are not a solution.

For me, your comment does not explain this. This comment suggest that this is something like change layergroups and reset view to adjust view to data. 

As data does not share a coordinate system, the only reason I see to use "spaces" is the fast switch between data points. If yes, then such spaces should have subspaces for different interpretations (labels, points, shapes).

Maybe this is  wording things but base on your description it is better for me to think about this as "states" not "spaces". It may be correct? 


My biggest wories about such feature are possible MemoryErrors. In PartSeg I need to add compressing labels data to reduce such errors when user store multiple segmentaions. But this of course slowdown state switch. 

To address my worries it will be nice to attach information about this feature usage to error report about memory problem for simplify understand user issues/bug reports.

**Comment by brisvag** [comment]
I think I wasn't very clear; when I say that they "do not share a coordinate system,", I don't mean simply that they are "in different places" or that they use different transforms; what I mean is that the "world" of layer 1 is not the same as the "world" of layer 2.

For example, imagine you collect two images (`i1` and `i2`) on you microscope, and load them as 2 layers. They are images from two different samples, whose feature you want to analyse, but that do not belong to the same plate/grid/support (that is, there is no specific (or relevant) spatial or temporal relationship between them).

You then do your processing, segmentation, and so on, and you end up with a bunch of layers. You can indeed group them in two groups (i.e: `i1+` and `i2+`). Now, what's the point of having the two groups *in the viewer*? You probably want to do something with the collective data; however, you don't want to *show* them at the same time. You might want to show them side by side for visual comparison, sometimes even overlap them for some quick morphological observation, but in most cases you will simply switch between showing one group, to showing another group.

In my case, this happens all the time, multiplied by *a lot*: there is almost no reason for them to be *on the canvas* together, but I want them somewhere "in the viewer" so I can access their data quickly (and this is what I was showing in the example above).  Also, note that this is worse in 3D because side-by-side view is basically useless until we have multi-canvas with separate cameras).

 I also don't want the layerlist to be a huge mess where I'm forced to scroll (or search) the layer that I'm looking at anyways. Even better, it would be nice if there was a shortcut to quickly switch between these spaces so one doesn't have to make visible/invisible specific layer groups every time.

So in short, while I agree that layergroups are still the way forward and will make interacting with the layerlist smarter and easier, I still think that some higher-level separation of "spaces" (or workspaces, or even "states" as you say, depending on how the uses uses it) would be a nice addition.

**Comment by Czaki** [comment]
> So in short, while I agree that layergroups are still the way forward and will make interacting with the layerlist smarter and easier, I still think that some higher-level separation of "spaces" (or workspaces, or even "states" as you say, depending on how the uses uses it) would be a nice addition.

I agree. :)

But before implementing this I would love to see multiple columns docking areas for a clear interface. 
And I think that It could be simply implemented as a plugin.

**Comment by alisterburt** [comment]
Counterpoint from discussion with @sofroniewn - if the spaces are really separate then why would the camera/dims be the same when changing spaces?

Because of that I'm now convinced that some ergonomic 'multi viewer' situation is actually what's most correct/flexible 

People have discussed 'tabbed' viewers in the past as an implementation of this - I think for large numbers of viewer states a searchable combobox containing viewer states to be switched in and out is the way to go (probably situated somewhere around the layer list)

Switching spaces in the same viewer and tabbed viewers aren't necessarily mutually exclusive - imagine if clicking a viewer state applied it on the current viewer, open in new tab or open in new window could be available through a context menu :)

**Comment by Czaki** [comment]
> I think for large numbers of viewer states a searchable combobox containing viewer states to be switched in and out is the way to go (probably situated somewhere around the layer list)

Searchable list widget or even searchable QTreeWidget. Searchable combobox is not a good choice for 100 entries with names generated by some reader.

**Comment by alisterburt** [comment]
@Czaki don't those two take up a lot more space?

**Comment by Czaki** [comment]
> don't those two take up a lot more space?

Yes, they take but are much more readable. And allow to easily understand the current state. 
 Maybe the proper solution is to have QComboBox by default and a larger widget after clicking some button.

**Comment by alisterburt** [comment]
@Czaki sounds great to me!

**Comment by Czaki** [comment]
And when you think about the size you need to remember that this is not only the size of a selector, but there needs to be an interface for managing the state. "Create space", "Delete space", "Rename space", "Update space" etc. And I'm not sure if such interface should be only in right-click menu

**Comment by brisvag** [comment]
> Counterpoint from discussion with @sofroniewn - if the spaces are really separate then why would the camera/dims be the same when changing spaces?
> 
> Because of that I'm now convinced that some ergonomic 'multi viewer' situation is actually what's most correct/flexible 
> 
> People have discussed 'tabbed' viewers in the past as an implementation of this - I think for large numbers of viewer states a searchable combobox containing viewer states to be switched in and out is the way to go (probably situated somewhere around the layer list)
> 
> Switching spaces in the same viewer and tabbed viewers aren't necessarily mutually exclusive - imagine if clicking a viewer state applied it on the current viewer, open in new tab or open in new window could be available through a context menu :)
> 
> 

Correct me if I'm wrong, but this just seems a difference of semantics and UX to me (tabbed vs dropdown vs tree), rather than of feature and implementation.

Indeed, camera and dims should not be the same, but reusing the model and restoring state is easier and just as effective until we can support real multi-camera/multidims/multicanvas.

**Comment by alisterburt** [comment]
@brisvag the visual stuff is definitely semantics - whether a space is just a set of layers or a more complete description of viewer state is pretty important though?

**Comment by brisvag** [comment]
Right - I think it's because I subconsciosuly was already assuming that viewer state should be preserved when going back to a space, but I guess I never explicitly stated that. So yeah, I think we agree that ultimately we want that :P

However, one thing I think is very important: this should be something *contained* by the `ViewerModel`. I don't want to juggle 100 different `Viewer` objects. I Imagine this being something that sits just above the layerlist, but below the viewer. In that case, I suppose a different `dims` and `camera` (and potentially `canvases`...) should exist for each space.

Alternatively (and this is easier to implemen wihtout big structural changes), we just make it quick way to save and then `serialize`/`deserialize` viewer states. This way we just have one viewer object to deal with.

**Comment by alisterburt** [comment]
Totally agree :) especially on the juggling multiple viewers!

yay, seems like we have some consensus

**Comment by sofroniewn** [comment]
> However, one thing I think is very important: this should be something contained by the ViewerModel. I don't want to juggle 100 different Viewer objects. I Imagine this being something that sits just above the layerlist, but below the viewer. In that case, I suppose a different dims and camera (and potentially canvases...) should exist for each space.

Just catching up here - this is where I might disagree. If dims & camera etc are changing then you really have different viewers as a viewer contains a dims and camera.

Some of this was discussed in #3956 and #3955.

We could have a top level API like `app` and then something like

```python
app.viewers # list of all viewers
app.viewer # current viewer model
app.window # the singleton window now lives on the app rather than the viewer
```

If you only have one viewer you'd just be working with `viewer = app.viewer` as before?

What does everyone think of this? It would probably be a 0.5.0 feature. I know @Carreau @jni @tlambert03 might have thoughts here too.

Is this such a major API change we might want to write a NAP?

**Comment by Czaki** [comment]
> Just catching up here - this is where I might disagree. If dims & camera etc are changing then you really have different viewers as a viewer contains a dims and camera.

No. You just could have to serialize/deserialize operations for state and use it to save/restore state with information about dims and camera positions. 

This of course will be slower than having multiple viewers with all data in GPU memory but maybe nicer for less powerful computers user.

**Comment by brisvag** [comment]
> Just catching up here - this is where I might disagree. If dims & camera etc are changing then you really have different viewers as a viewer contains a dims and camera.

This is the way I see it: if you think of the viewer + dims + camera as your photography room/light setup/camera, layers are the objects and people that you're imaging, and spaces encode the *way* you arrange those subjects in your room (and which ones). At any times you should eb able to pick up an object from your storage and stick it in the room. You should also be able to reconstruct the room as it was two days ago. Anopther viewer would juyst be a *completely new* room, that has no shared storage or history with the other one.

> Is this such a major API change we might want to write a NAP?

Either way, probably yes :P As soon as we have collected enough thoughts, I might do that!

**Comment by Czaki** [comment]
Adding serialization/deserialization layers spaces will allow to save it in one viewer and load in another.

**Comment by andy-sweet** [comment]
Trying to catch up here, so I'm going to try to summarize what I understand from the discussion above. Overall, I understand the motivation for the feature and think there are some solid ideas already.

> This is the way I see it: if you think of the viewer + dims + camera as your photography room/light setup/camera, layers are the objects and people that you're imaging, and spaces encode the _way_ you arrange those subjects in your room (and which ones).

This is effectively how I think about slicing/sampling the layers too. We're using the current state of (dims, camera, layerlist) as input and generating the sliced/sampled data + other stuff that we want to present in the canvas (e.g. that we send to vispy).

In the feature described here we have multiple sets of (dims, camera, layerlist) that we can switch between and one layer only belongs to one set. In what I imagine multi-canvas to be, we can present many of those sets of (dims, camera, layerlist) simultaneously and a layer can be presented in multiple canvases.

So I think this feature makes a nice precursor to multi-viewer/canvas, as it supports multiple viewers in the background, but doesn't work out all the details of how to present all of those in the GUI simultaneously. I think it also doesn't depend on removing the slice/sample state from the layers (because one layer belongs to exactly one space), whereas multi-canvas probably does.

> However, one thing I think is very important: this should be something _contained_ by the `ViewerModel`. I don't want to juggle 100 different `Viewer` objects.

Currently the `Viewer` is the `ViewerModel` due to inheritance. The `Viewer` also contains `Window` and thus `QtViewer`, some of which will need to at least be aware of the current space (e.g. `QtDims` needs to update to reflect the current space's `Dims`). Again, I think multi-canvas/viewer is a harder problem that requires major surgery on `QtViewer` because some widgets may need to be per-viewer (e.g. `QtDims`) whereas others probably don't (e.g. `QtConsole`).

**Comment by sofroniewn** [comment]
> No. You just could have to serialize/deserialize operations for state and use it to save/restore state with information about dims and camera positions.

I think we absolutely want to support full serialization of the whole viewer state to disk, but I see this as more akin to saving and loading files.

> This of course will be slower than having multiple viewers with all data in GPU memory but maybe nicer for less powerful computers user.

I can see how memory limits could become a concern, but maybe we can address them by not sending over data from viewers that you're not looking at.

> > Is this such a major API change we might want to write a NAP?
>
> Either way, probably yes :P As soon as we have collected enough thoughts, I might do that!

Great, I think what I'd like to see are API proposals and architecture diagrams here.

I'm still most excited about the API I proposed above https://github.com/napari/napari/issues/4419#issuecomment-1129443846

**Comment by alisterburt** [comment]
> We could have a top level API like `app` and then something like
> 
> ```python
> app.viewers # list of all viewers
> app.viewer # current viewer model
> app.window # the singleton window now lives on the app rather than the viewer
> ```
> 
> If you only have one viewer you'd just be working with `viewer = app.viewer` as before?
> 
> What does everyone think of this? It would probably be a 0.5.0 feature. I know @Carreau @jni @tlambert03 might have thoughts here too.

I really like this. One thing I would like is to maintain our current `Viewer` API as the primary creation API in script world and having an 'accessor' for the current `app` if it is needed

something like

```python
import napari

viewer = napari.Viewer()
app = napari.get_app()
```

rather than

```python
import napari

app = napari.Application()
viewer = app.viewer
```

these are details that can be worked out as we go, overall I'm super excited about this!!

**Comment by tlambert03** [comment]
I haven't read everything carefully... so forgive me if this was covered,  but:
- my first thoughts are with @jni: "isn't this just multiple viewers?"
   - it seems like @brisvag's primary comment in response to @jni saying "isn't this just multiple viewers?" was "yes, but I have _hundreds_" of them.  Which doesn't necessarily mean that thinking of this as multiple viewers is incorrect, but just that we need to have better ways of hiding/showing/grouping `Viewers`.  Something like a `ViewerList`, akin to a layerlist, where each viewer can have some visibility and other state that determines how computationally "active" it is.

- my second thought is that we have generally _under_-utilized the `LayerList` itself as a swappable object.  We think of a viewer as having one permanent/indelible layerlist... but there's no conceptual reason that you shouldn't be able to change the layerlist with `viewer.layers = other_space`.  (certainly, we'd have a lot of cleanup to do to make this doable, since we've generally been rather lazy about _disconnecting_ things... which is why things like #3184 are so hard to do).  But really, the isolation is mostly in place to let a viewer change out its layerlist.  The viewer is mostly just a view onto a layerlist model.  So inasmuch as we _can't_ do this it's kind of an oversight/bug.
- in the case of a swappable layerlist, one then has to ask, what _is_ the viewer beyond a set of layers, which comes back to @sofroniewn's comments about dims and cameras.  So, in the case of spaces here, is there really anything else about the viewer (beyond just the layerlist) that needs/wants to be synchronized across spaces?  Do you _want_ the camera to stay constant when you switch between spaces?  And if not, doesn't that just bring us back to multiple viewers?

- `app.viewers` ... we do essentially have this already (it's `napari.Viewer._instances`).  I'm not at all opposed to another level of abstraction by creating an `app` API, but I sort of see that as a separate issue (i.e. until we have something _other_ than a set of `Viewers` that we want to track, discussion of that API is a bit orthogonal to this here).

(side note: I also do think that the side question/problem of reader plugins not being able to create multiple viewers/layerlists is a bit interesting, certainly as datasets get massive...)

**Comment by Czaki** [comment]
Part of the layers should be sharable between spaces/states. Something like data with different interpretations.

**Comment by tlambert03** [comment]
> Something like data with different interpretations.

currently the layer _is_ the interpretation of the data... (i.e. it's a tight coupling between `data` and some display properties) but once we abstract out the concept of a "data source", then it would be much easier to accomplish this, as multiple layers (i.e. "interpretations") could share the same data source.

**Comment by alisterburt** [comment]
> Which doesn't necessarily mean that thinking of this as multiple viewers is incorrect, but just that we need to have better ways of hiding/showing/grouping Viewers. Something like a ViewerList, akin to a layerlist, where each viewer can have some visibility and other state that determines how computationally "active" it is.

Perfect summary! It feels like we've reached consensus here on 'spaces' being multiple viewers. 

> we have generally under-utilized the LayerList itself as a swappable object.

> in the case of a swappable layerlist, one then has to ask, what is the viewer beyond a set of layers, which comes back to @sofroniewn's comments about dims and cameras.

I agree that being able to swap the layerlist in and out would be great and a good example of isolation being well done but I can't see when this would be necessary/more appropriate than switching viewers in the case where we have great viewer switching functionality - do you agree here?

I think being to both switch the layerlist and the viewer will likely create a situation where some plugin developers use the multiple viewers API, some use multiple layerlists... It feels important to handle this idea of  'data in multiple spaces'  uniformly within the plugin ecosystem. I still *really* like the idea of batching analysis workflows across multiple spaces (viewers) from https://github.com/napari/napari/issues/4419#issuecomment-1126353589  

> i.e. until we have something other than a set of Viewers that we want to track, discussion of that API is a bit orthogonal to this here)

Agreed, the obvious candidate for something which would best be stored at `app` level is data sources accessible in different viewers but it's good to know that this can be thought of separately

> (side note: I also do think that the side question/problem of reader plugins not being able to create multiple viewers/layerlists is a bit interesting, certainly as datasets get massive...)

Yeah, this is absolutely the motivating case for some incarnation of 'spaces' (multiple viewers) and is one reason why reader plugins haven't worked for us (you previously mentioned handling hdf5 being limited by similar issues)

**Comment by tlambert03** [comment]
> I agree that being able to swap the layerlist in and out would be great and a good example of isolation being well done but I can't see when this would be necessary/more appropriate than switching viewers in the case where we have great viewer switching functionality - do you agree here?

It would be more appropriate when (and only when) you wanted to keep the _other_ items in the `ViewerModel` (i.e. `axes`, `camera`, `dims`, `grid`, etc...) constant, but _only_ change the set of layers. (not sure if that's the case here... but that's when)

> I still really like the idea of batching analysis workflows across multiple spaces (viewers) from https://github.com/napari/napari/issues/4419#issuecomment-1126353589

In the past, @sofroniewn and I have enjoyed thinking about the `LayerList` as the (view-independent) object that might serve as a good unit to pass around in workflows.  (I think we should try to keep gui objects like the viewer out of workflow discussion as much as possible)

**Comment by alisterburt** [comment]
> It would be more appropriate when (and only when) you wanted to keep the _other_ items in the `ViewerModel` (i.e. `axes`, `camera`, `dims`, `grid`, etc...) constant, but _only_ change the set of layers. (not sure if that's the case here... but that's when)

I'm with you, this has got my mind whirring and I think there's a useful nugget here. It feels like we need to support
- have a single viewer with relatively constant depiction (i.e. only swap data in and out of a single viewer/set of layers)
- have multiple viewers with different depictions of (potentially different, or not) data

Once we have the separate `DataSource` concept, both become much easier

> In the past, @sofroniewn and I have enjoyed thinking about the `LayerList` as the (view-independent) object that might serve as a good unit to pass around in workflows. (I think we should try to keep gui objects like the viewer out of workflow discussion as much as possible)

This is interesting and really good to keep in mind - it feels somewhat incompatible with the separation of `DataSource`

**Comment by sofroniewn** [comment]
> > In the past, @sofroniewn and I have enjoyed thinking about the LayerList as the (view-independent) object that might serve as a good unit to pass around in workflows. (I think we should try to keep gui objects like the viewer out of workflow discussion as much as possible)
>
> This is interesting and really good to keep in mind - it feels somewhat incompatible with the separation of DataSource

If we end up with a list of DataSources that can be placed into different layers then that might become a more natural unit to pass around workflows - I think key idea was we wanted what get's passed around workflows to view independent. This is maybe a side discussion though, best for another issue.


> my second thought is that we have generally under-utilized the LayerList itself as a swappable object ... The viewer is mostly just a view onto a layerlist model. So inasmuch as we can't do this it's kind of an oversight/bug.

Completely agree with this - and if this isn't working this is something that we should fix regardless of the outcome of this discussion.

> app.viewers ... we do essentially have this already (it's napari.Viewer._instances). I'm not at all opposed to another level of abstraction by creating an app API, but I sort of see that as a separate issue (i.e. until we have something other than a set of Viewers that we want to track, discussion of that API is a bit orthogonal to this here).

I guess for me this is just a super concrete way to understand and expose the usage through a public API. While it might not be strictly necessary I find it helpful as it can give a clear publicly documentable way for folks to move between different viewers and elevates the behavior to being first class. I wonder what @jni thinks

> I really like this. One thing I would like is to maintain our current Viewer API as the primary creation API in script world and having an 'accessor' for the current app if it is needed

I completely agree with this @alisterburt.

-------

I think this thread is getting quite long now - lots of great ideas and great discussion, but I want to make sure that it doesn't fizzle out / become too hard for folks to follow. I wonder if @brisvag and others have enough input here to begin writing a NAP with an implementation proposal for whatever they think is best. If not, then we can keep discussing here more too

**Comment by brisvag** [comment]
I think my main initial confusion about this discussion boils down to this: what do we mean by a viewer?
- A literal `ViewerModel` with all the stucck that comes with it right now, a completely independent object that can spawn its own GUI (either as a separate windows, or as a tab, or whatelse)? 
- Or just an object that "serializes" a layerlist and info about axes, camera, and dims? (I might be using the serialize term improperly here, judging by some of the answers above; I simply mean "describe and remember some info so they can be retrieved and reconstructed as needed") 

If it is the *former*, then I disagree, and that's my fear when I talk about "juggling a million objects". Will the console belong to a viewer, then, or to the app?

If it is the latter (as it seems, from the last few comments), then a viewer *is* a space. I'm fine with that, I'm not particularly attached to the terminology; I found it confusing because *viewer* is very much tied to the *window*, in my mind. You say "open the napari viewer", not "open the napari window, which is displaying the contents of a viewer" :P

I'm pretty sure everyone seems to be leaning towards the latter (correct me if I'm wrong!). I'll try to summarize some thoughts:

- while swapping layerlists in and out of the viewer would be fine for many cases (In fact, I *did* this before, with some ugly workarounds to the current bugs due to lack of disconnections mentioned by @tlambert03), the ability to switch *viewer* state as well seems pretty important, so something *smarter* seems useful
- `DataSource` will definitely be useful to make this able to look at *different interpretations* of the data; in the meantime, I don't think this possibility precludes any of the ideas we proposed, so that's good.
- I like the `app` idea, but I agree with @alisterburt that the *primary* interface should probably be the Viewer. In general, I think that the *key* issue to solve here is to make it *easy* to find the viewer/data you want to display. (searchable dropdown, zoommable graph, voice-activated assistant...)

I will try to write this stuff up in a NAP soon. Thanks everyone for the input, very cool discussion :)

**Comment by tlambert03** [comment]
> Will the console belong to a viewer, then, or to the app?

I'm not sure bringing in the scope of the console in to this discussion is terribly illuminating.  It too is an independent entity.  Its scope is complicated (neither the viewer nor the app)... and if we wanted to, we could probably make it appear to have a scope/context that was more or less limited.  but again, I see that also as an orthogonal discussion.

> what do we mean by a viewer?

well... I think your first description is the `Viewer` and the second description is the `ViewerModel` :joy: ...   `Viewer` is simply a `ViewerModel` with a window.  Though perhaps it's semantic: the `Viewer` is really not much of anything.  It's a composition of a `ViewerModel` and a `Window` (where the `Window` is a view on to the `ViewerModel`).

> the ability to switch viewer state as well seems pretty important, so something smarter seems useful

this comes back to my question above of "is there really anything else about the viewer (beyond just the layerlist) that needs/wants to be synchronized across spaces?"  If you have a set of things _beyond_ the set of layers that you want to swap in/out (like camera, dims, etc...) , then yeah, we need something like a full `ViewerList` (beyond just a swappable `LayerList`).

**Comment by brisvag** [comment]
This is now a NAP!

- #4684

**Comment by brisvag** [comment]
Closing this, since the discussion has been moved to the NAP.
